nifi-extension-bundles/nifi-standard-services/nifi-oauth2-provider-bundle/nifi-oauth2-provider-service/src/main/java/org/apache/nifi/oauth2/JWTBearerOAuth2AccessTokenProvider.java

```diff
@@ -0,0 +1,677 @@
+            claimsSetBuilder.jwtID(jti);
+        }
+
+        customClaims.forEach(claimsSetBuilder::claim);
+
+        JWSHeader.Builder headerBuilder = new JWSHeader.Builder(JWSAlgorithm.parse(algorithmName));
+
+        if (headerType) {
+            headerBuilder.type(JOSEObjectType.JWT);
+        }
+
+        if (kid != null) {
+            headerBuilder.keyID(kid);
+        }
+
+        if (headerX5T) {
+            try {
+                final String alias = keyManager.chooseClientAlias(new String[] {"RSA"}, null, null);
+                if (alias == null) {
+                    getLogger().warn("No key alias found, setting x5t header to null.");
+                } else {
+                    final PrivateKey privateKey = keyManager.getPrivateKey(alias);
+                    if (privateKey == null) {
+                        getLogger().warn("No private key found, setting x5t header to null.");
+                    } else {
+                        final X509Certificate[] certificates = keyManager.getCertificateChain(alias);
+                        if (certificates == null || certificates.length == 0) {
+                            getLogger().warn("No certificates found");
+                        } else {
+                            final MessageDigest messageDigest = MessageDigest.getInstance("SHA-256");
+                            final byte[] bytes = messageDigest.digest(certificates[0].getEncoded());
+                            String url = Base64.getEncoder().encodeToString(bytes);
+                            headerBuilder.x509CertSHA256Thumbprint(new Base64URL(url));
+                        }
+                    }
+                }
+            } catch (Exception e) {
+                getLogger().warn("Failed to set x5t header, setting to null.", e);
+            }
+        }
+
+        Map<String, String> formParams = new HashMap<>();
+        formParams.put("grant_type", grantType);
+        formParams.put(assertion, getAssertion(headerBuilder.build(), claimsSetBuilder.build()));
+        formParams.putAll(this.formParams);
+
+        requestTokenEndpoint(formParams);
+    }
+
+    protected void requestTokenEndpoint(Map<String, String> formParams) throws URISyntaxException {
+        HttpRequestHeadersSpec request = webClientService
+                .post()
+                .uri(new URI(tokenEndpoint))
+                .header("Accept", ContentType.APPLICATION_JSON.getType())
+                .header("Content-Type", ContentType.APPLICATION_URLENCODED.getType())
+                .body(formParams.entrySet()
+                        .stream()
+                        .map(param -> param.getKey() + "=" + param.getValue())
+                        .collect(Collectors.joining("&")));
+
+        try (final HttpResponseEntity response = request.retrieve()) {
+            if (response.statusCode() != 200) {
+                final String message = "Failed to retrieve access details. Status code: %s, response body: %s".formatted(
```

Comment:
Recommend adjusting the formatting
```suggestion
                final String message = "Failed to retrieve Access Token: HTTP %s with Response [%s]".formatted(
```